### PR TITLE
sql: improve performance of schema change operations

### PR DESCRIPTION
**Pull request by vivekmenezes**
This change improves performance by scanning fewer keys
and eliminating obvious redundant operations.

Adjusted the chunking constants to be more appropriate.
A schema change is still affecting the throughput of OLTP
operations.

Added context to various function calls, and log to traces where
appropriate.
#9149

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/9798)

### CODE REVIEW

**Comment by tamird** [comment]
Why does this all need to be in one commit? Seems like there's the context plumbing for logging, and some other refactoring, and the two are quite separate and could use their own detailed commit messages.

---

Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[sql/backfill.go, line 35 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTUmFv3uFKeFpftvQAZ:-KTUmFv4YydL2jTn7PYd:b-x7ss7z) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a6f7315f623b80c49cf280d3d749f274aec76c4/sql/backfill.go#L35)):*
> ```Go
> 	// operation chunk involved.
> 
> 	// ColumnTruncateAndBackfillChunkSize is the maximum number of rows of
> ```

s/of rows//

---

*[sql/backfill.go, line 245 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTUmcMQDGJ-_4jRxetf:-KTUmcMR4q5bemWZDCfd:b-x8ptv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a6f7315f623b80c49cf280d3d749f274aec76c4/sql/backfill.go#L245)):*
> ```Go
> 
> 		// Run through the entire table key space adding and deleting columns.
> 		for row, done := 0, false; !done; row += ColumnTruncateAndBackfillChunkSize {
> ```

let's not spread knowledge of this constant around. if you're going to use it here, make it an argument to `truncateAndBackfillColumnsChunk` so that function doesn't have to know about it.

---

*[sql/backfill.go, line 329 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTUmz4lQK5AoL7tInxl:-KTUmz4lQK5AoL7tInxm:bbnhteo) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a6f7315f623b80c49cf280d3d749f274aec76c4/sql/backfill.go#L329)):*
> ```Go
> 			_, valNeededForCol[i] = ru.fetchColIDtoRowIndex[tableDesc.Columns[i].ID]
> 		}
> 		err = rf.Init(tableDesc, colIDtoRowIndex, &tableDesc.PrimaryIndex, false, false,
> ```

while you're here, mind reducing the scope of these `err`s?

---

*[sql/backfill.go, line 342 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTUn3R7f2g4ZlpTdBEd:-KTUn3R7f2g4ZlpTdBEe:bz2j0bo) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a6f7315f623b80c49cf280d3d749f274aec76c4/sql/backfill.go#L342)):*
> ```Go
> 		oldValues := make(parser.DTuple, len(ru.fetchCols))
> 		updateValues := make(parser.DTuple, len(updateCols))
> 		var nonNullViolationColumn string
> ```

s/Column/ColumName/

---

*[sql/backfill.go, line 353 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTUnXvgWYKjRT01w5uc:-KTUnXvgWYKjRT01w5ud:b-io87ao) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a6f7315f623b80c49cf280d3d749f274aec76c4/sql/backfill.go#L353)):*
> ```Go
> 			}
> 			if !col.Nullable && updateValues[j].Compare(parser.DNull) == 0 {
> 				nonNullViolationColumn = col.Name
> ```

this refactor seems to saving O(added-indexes) iterations per row, is that right? is this meaningful? where is the benchmark?

---

*[sql/backfill.go, line 360 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTUnAjWTG_rOtH0Rkqp:-KTUnAjWTG_rOtH0Rkqq:b-qna6e5) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a6f7315f623b80c49cf280d3d749f274aec76c4/sql/backfill.go#L360)):*
> ```Go
> 		}
> 
> 		writeBatch := &client.Batch{}
> ```

txn.NewBatch

---

*[sql/backfill.go, line 377 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTUnrRsXw2cuTTedyXs:-KTUnrRsXw2cuTTedyXt:b-ebzxfd) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a6f7315f623b80c49cf280d3d749f274aec76c4/sql/backfill.go#L377)):*
> ```Go
> 
> 			copy(oldValues, row)
> 			if rowLength != len(row) {
> ```

this code needs some comments

---

*[sql/backfill.go, line 515 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTUoajl2B46z7TTFkSB:-KTUoajl2B46z7TTFkSC:b-h5eh5v) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a6f7315f623b80c49cf280d3d749f274aec76c4/sql/backfill.go#L515)):*
> ```Go
> 		scan.desc = *tableDesc
> 		scan.spans = []roachpb.Span{sp}
> 		scan.SetLimitHint(IndexBackfillChunkSize, false)
> ```

ditto, seems like this wants to be an argument rather than use the constant directly

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-:-KTUod_TJ2iItEESFwOA:b-41nij6)*

**Comment by vivekmenezes** [comment]
I've moved the context plumbing into a different commit.

---

Review status: 3 of 5 files reviewed at latest revision, 8 unresolved discussions.

---

*[sql/backfill.go, line 35 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTUmFv3uFKeFpftvQAZ:-KTiosXOmoTT9IK67ZG6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a6f7315f623b80c49cf280d3d749f274aec76c4/sql/backfill.go#L35)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/of rows//
</details>
Done.

---

*[sql/backfill.go, line 245 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTUmcMQDGJ-_4jRxetf:-KTiotyFChXUAklKGbA-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a6f7315f623b80c49cf280d3d749f274aec76c4/sql/backfill.go#L245)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> let's not spread knowledge of this constant around. if you're going to use it here, make it an argument to `truncateAndBackfillColumnsChunk` so that function doesn't have to know about it.
</details>
Done.

---

*[sql/backfill.go, line 329 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTUmz4lQK5AoL7tInxl:-KTiovNDPtLnji7Sg38Z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a6f7315f623b80c49cf280d3d749f274aec76c4/sql/backfill.go#L329)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> while you're here, mind reducing the scope of these `err`s?
</details>
Done.

---

*[sql/backfill.go, line 342 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTUn3R7f2g4ZlpTdBEd:-KTimwgMkEE5ugm3cLDb:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a6f7315f623b80c49cf280d3d749f274aec76c4/sql/backfill.go#L342)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/Column/ColumName/
</details>
done

---

*[sql/backfill.go, line 353 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTUnXvgWYKjRT01w5uc:-KTip-hnPOR1blHKMVu8:b-cpyr9f) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a6f7315f623b80c49cf280d3d749f274aec76c4/sql/backfill.go#L353)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this refactor seems to saving O(added-indexes) iterations per row, is that right? is this meaningful? where is the benchmark?
</details>
This takes out this code from the for loop below. O(chunkSize) -> O(1) 

---

*[sql/backfill.go, line 360 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTUnAjWTG_rOtH0Rkqp:-KTipTEK6LW8yjEyDkWJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a6f7315f623b80c49cf280d3d749f274aec76c4/sql/backfill.go#L360)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> txn.NewBatch
</details>
Done.

---

*[sql/backfill.go, line 377 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTUnrRsXw2cuTTedyXs:-KTiqqZ1xOVsfab7jXbB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a6f7315f623b80c49cf280d3d749f274aec76c4/sql/backfill.go#L377)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this code needs some comments
</details>
Done.

---

*[sql/backfill.go, line 515 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTUoajl2B46z7TTFkSB:-KTiqsWgzDz8HPG68PdJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a6f7315f623b80c49cf280d3d749f274aec76c4/sql/backfill.go#L515)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ditto, seems like this wants to be an argument rather than use the constant directly
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-:-KTilvahLxrS-2GcSngu:boiueo2)*

**Comment by tamird** [comment]
Thanks. Note that it seems that this context is not annotate with anything interesting, so it seems that this goes against https://github.com/cockroachdb/cockroach/issues/9807 for the sake of convenience. I think we just need to plumb these contexts instead of putting them on `schemaChanger`.

---

Reviewed 2 of 2 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[sql/backfill.go, line 245 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTUmcMQDGJ-_4jRxetf:-KTjMH-_UINXqxRLG6cY:b-xnkj0r) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a6f7315f623b80c49cf280d3d749f274aec76c4/sql/backfill.go#L245)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> Done.
</details>
Now that this constant is only used here, perhaps it should be scoped to the function. `const chunkSize = ....`

---

*[sql/backfill.go, line 353 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTUnXvgWYKjRT01w5uc:-KTjMUydYFBcZY6RWTA1:bt4the7) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a6f7315f623b80c49cf280d3d749f274aec76c4/sql/backfill.go#L353)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> This takes out this code from the for loop below. O(chunkSize) -> O(1) 
</details>
Sure. It would still be good to have a benchmark to justify the refactor.

---

*[sql/backfill.go, line 377 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTUnrRsXw2cuTTedyXs:-KTjMfPdSpARNow33_O1:bwj156f) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a6f7315f623b80c49cf280d3d749f274aec76c4/sql/backfill.go#L377)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> Done.
</details>
Sorry, I meant that a comment is needed about the `len(row)` shenanigans.

---

*[sql/backfill.go, line 39 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTjMAMus9LUg3j_V6-6:-KTjMAMus9LUg3j_V6-7:b-kyfvkw) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5464f3fb7671ebc4c160926919b9ae7efc82ae2/sql/backfill.go#L39)):*
> ```Go
> 	ColumnTruncateAndBackfillChunkSize = 200
> 
> 	// IndexTruncateChunkSize is the maximum number of rows processed per
> ```

aren't these keys, not rows?

---

*[sql/backfill.go, line 45 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTjMCL4cLMz-Fs9T5Ve:-KTjMCL4cLMz-Fs9T5Vf:b-kyfvkw) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5464f3fb7671ebc4c160926919b9ae7efc82ae2/sql/backfill.go#L45)):*
> ```Go
> 	IndexTruncateChunkSize = 600
> 
> 	// IndexBackfillChunkSize is the maximum number of rows processed per
> ```

aren't these keys, not rows?

---

*[sql/backfill.go, line 377 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTjMdk0GRKUoZgzfTVM:-KTjMdk1kvYy-ZKEX0t8:b-gfoc27) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5464f3fb7671ebc4c160926919b9ae7efc82ae2/sql/backfill.go#L377)):*
> ```Go
> 
> 			copy(oldValues, row)
> 			// Update oldValues with NULL values where values were'nt found.
> ```

"were'nt" typo

---

*[sql/backfill.go, line 408 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTjMn4erTY_7vXQ8Akv:-KTjMn4erTY_7vXQ8Akw:b-xnkj0r) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5464f3fb7671ebc4c160926919b9ae7efc82ae2/sql/backfill.go#L408)):*
> ```Go
> 	lease *sqlbase.TableDescriptor_SchemaChangeLease, dropped []sqlbase.IndexDescriptor,
> ) error {
> 	chunkSize := int64(IndexTruncateChunkSize)
> ```

Now that this constant is only used here, perhaps it should be scoped to the function. `const chunkSize = ....`

---

*[sql/backfill.go, line 468 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTjMpyihN-bZo3P7tO0:-KTjMpyihN-bZo3P7tO1:b-xnkj0r) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5464f3fb7671ebc4c160926919b9ae7efc82ae2/sql/backfill.go#L468)):*
> ```Go
> 
> 	// Backfill the index entries for all the rows.
> 	chunkSize := int64(IndexBackfillChunkSize)
> ```

Now that this constant is only used here, perhaps it should be scoped to the function. `const chunkSize = ....`

---

*[sql/truncate.go, line 127 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTjMwSpREBXrjhSzH9M:-KTjMwSpREBXrjhSzH9N:b-xnkj0r) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5464f3fb7671ebc4c160926919b9ae7efc82ae2/sql/truncate.go#L127)):*
> ```Go
> ) error {
> 	var resume roachpb.Span
> 	for row, done := 0, false; !done; row += TableTruncateChunkSize {
> ```

Now that this constant is only used here, perhaps it should be scoped to the function. `const chunkSize = ....`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-:-KTjLpX0bq9288O-fTrI:b-q75a88)*

**Comment by RaduBerinde** [comment]
The context situation in this area of SQL is not yet figured out. I think it will take a bit of thinking of where and how we need to plumb contexts correctly here. This is distracting from the main point of the change, so I think it's better to leave the contexts as TODO for now (unless there was something that you needed fixed up for debugging). I go over them as part of #7992; better to wait until we get a handle of the new model proposed in #9807 anyway.

---

Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-:-KTjm1zIz3QGjnx7jZB6:b2opglc)*

**Comment by vivekmenezes** [comment]
I'm okay with removing this commit because I'm not really using it. Thanks

**Comment by vivekmenezes** [comment]
Review status: 1 of 3 files reviewed at latest revision, 9 unresolved discussions.

---

*[sql/backfill.go, line 245 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTUmcMQDGJ-_4jRxetf:-KTjx5_VoCMM0lw-VNq2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a6f7315f623b80c49cf280d3d749f274aec76c4/sql/backfill.go#L245)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Now that this constant is only used here, perhaps it should be scoped to the function. `const chunkSize = ....`
</details>
Done.

---

*[sql/backfill.go, line 353 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTUnXvgWYKjRT01w5uc:-KTjx7i58z0t-ydq_f8i:b-1w2ywa) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a6f7315f623b80c49cf280d3d749f274aec76c4/sql/backfill.go#L353)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Sure. It would still be good to have a benchmark to justify the refactor.
</details>
well that's going to involve a lot more and I'm not adding a benchmark in this PR. I'm just taking out code that is running within a loop and placing it outside the loop. 

I've improved this further to share the code across many transactions. So now when updating a massive table the defaults are 
evaluated only once.

---

*[sql/backfill.go, line 377 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTUnrRsXw2cuTTedyXs:-KTjzH-beFAaI2upzbnI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a6f7315f623b80c49cf280d3d749f274aec76c4/sql/backfill.go#L377)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Sorry, I meant that a comment is needed about the `len(row)` shenanigans.
</details>
Done.

---

*[sql/backfill.go, line 39 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTjMAMus9LUg3j_V6-6:-KTjwVjoGEyGGtoRC3vX:b-lltkj3) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5464f3fb7671ebc4c160926919b9ae7efc82ae2/sql/backfill.go#L39)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> aren't these keys, not rows?
</details>
removed reference to rows/keys and replaced them with column or index .

---

*[sql/backfill.go, line 45 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTjMCL4cLMz-Fs9T5Ve:-KTjwfLKJpEquU1b2GM1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5464f3fb7671ebc4c160926919b9ae7efc82ae2/sql/backfill.go#L45)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> aren't these keys, not rows?
</details>
Done.

---

*[sql/backfill.go, line 377 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTjMdk0GRKUoZgzfTVM:-KTjzHPhmhwU36W831C_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5464f3fb7671ebc4c160926919b9ae7efc82ae2/sql/backfill.go#L377)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> "were'nt" typo
</details>
Done.

---

*[sql/backfill.go, line 408 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTjMn4erTY_7vXQ8Akv:-KTjzMVO5lViCnFtZUCD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5464f3fb7671ebc4c160926919b9ae7efc82ae2/sql/backfill.go#L408)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Now that this constant is only used here, perhaps it should be scoped to the function. `const chunkSize = ....`
</details>
Done.

---

*[sql/backfill.go, line 468 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTjMpyihN-bZo3P7tO0:-KTjzQdQ7oKEDqShan8Q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5464f3fb7671ebc4c160926919b9ae7efc82ae2/sql/backfill.go#L468)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Now that this constant is only used here, perhaps it should be scoped to the function. `const chunkSize = ....`
</details>
Done.

---

*[sql/truncate.go, line 127 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTjMwSpREBXrjhSzH9M:-KTjzWOtclM_fFUcS6Gu:b-hfpv0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5464f3fb7671ebc4c160926919b9ae7efc82ae2/sql/truncate.go#L127)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Now that this constant is only used here, perhaps it should be scoped to the function. `const chunkSize = ....`
</details>
removed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798)*

**Comment by tamird** [comment]
Reviewed 3 of 4 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[sql/backfill.go, line 245 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTUmcMQDGJ-_4jRxetf:-KTkC_BuNKYENAAtTp5o:b3girch) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a6f7315f623b80c49cf280d3d749f274aec76c4/sql/backfill.go#L245)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> Done.
</details>
Sorry, this is not what I meant - my suggestion was to move the actual definition of the constant here, removing the exported constant altogether.

---

*[sql/backfill.go, line 353 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTUnXvgWYKjRT01w5uc:-KTkD-uKgICjmej48wF9:bgtnu5j) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a6f7315f623b80c49cf280d3d749f274aec76c4/sql/backfill.go#L353)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> well that's going to involve a lot more and I'm not adding a benchmark in this PR. I'm just taking out code that is running within a loop and placing it outside the loop. 
> 
> I've improved this further to share the code across many transactions. So now when updating a massive table the defaults are 
> evaluated only once.
</details>
@paperstreet should review this, I guess.

---

*[sql/backfill.go, line 408 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTjMn4erTY_7vXQ8Akv:-KTkCpaQUicPqynFno_8:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5464f3fb7671ebc4c160926919b9ae7efc82ae2/sql/backfill.go#L408)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> Done.
</details>
ditto

---

*[sql/backfill.go, line 468 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTjMpyihN-bZo3P7tO0:-KTkCr4OEhRRieWW5VLX:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5464f3fb7671ebc4c160926919b9ae7efc82ae2/sql/backfill.go#L468)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> Done.
</details>
ditto

---

*[sql/truncate.go, line 127 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTjMwSpREBXrjhSzH9M:-KTkCrh_ylwL1rqyLtII:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5464f3fb7671ebc4c160926919b9ae7efc82ae2/sql/truncate.go#L127)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> removed
</details>
ditto

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798)*

**Comment by vivekmenezes** [comment]
Review status: 1 of 3 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[sql/backfill.go, line 245 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTUmcMQDGJ-_4jRxetf:-KTkI7_tI1grY_mJFprK:brg8lkq) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a6f7315f623b80c49cf280d3d749f274aec76c4/sql/backfill.go#L245)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Sorry, this is not what I meant - my suggestion was to move the actual definition of the constant here, removing the exported constant altogether.
</details>
I like all the constants in one place, so I'm leaving them as is; besides there is a TODO above them to eliminate them eventually. I've unexported them.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798)*

**Comment by paperstreet** [comment]
Review status: 1 of 3 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[sql/backfill.go, line 354 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTpQIBddOjdAxAmIE2F:-KTpQIBddOjdAxAmIE2G:b-5y0xci) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f56246c42090d4c4023bc63bac40dbea2f04f4f/sql/backfill.go#L354)):*
> ```Go
> 			_, valNeededForCol[i] = ru.fetchColIDtoRowIndex[tableDesc.Columns[i].ID]
> 		}
> 		if err := rf.Init(tableDesc, colIDtoRowIndex, &tableDesc.PrimaryIndex, false, false,
> ```

these were like this because it's hard to follow where the ()s and the {}s start and end. either revert or wrap `); err != {` down onto the next line and align it with the `if`

---

*[sql/backfill.go, line 425 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTpPaQIId7vtFJMDhma:-KTpPaQIId7vtFJMDhmb:b-2v0eeh) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f56246c42090d4c4023bc63bac40dbea2f04f4f/sql/backfill.go#L425)):*
> ```Go
> 
> 			resumeAt := resume
> 			log.Infof(context.TODO(), "drop index (%d, %d) at row: %d, span: %s",
> ```

seems like these should be wrapped in log.V(something)? here and elsewhere

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798)*

**Comment by vivekmenezes** [comment]
Review status: 0 of 3 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[sql/backfill.go, line 354 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTpQIBddOjdAxAmIE2F:-KTqPDiTYpPJtgE35xSo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f56246c42090d4c4023bc63bac40dbea2f04f4f/sql/backfill.go#L354)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> these were like this because it's hard to follow where the ()s and the {}s start and end. either revert or wrap `); err != {` down onto the next line and align it with the `if`
</details>
Done.

---

*[sql/backfill.go, line 425 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-KTpPaQIId7vtFJMDhma:-KTqPqNDSe_24UT3XUtj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f56246c42090d4c4023bc63bac40dbea2f04f4f/sql/backfill.go#L425)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> seems like these should be wrapped in log.V(something)? here and elsewhere
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798)*

**Comment by paperstreet** [comment]
:lgtm:

---

Review status: 0 of 3 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9798#-:-KTtQGuZ4fYwQ_cN45Sn:bnfp4nl)*
